lib/Ix/Context/WithAccount.pm

```diff
@@ -34,15 +36,92 @@ has root_context => (
   ) ],
 );
 
-sub process_request ($self, $calls) {
-  $self->processor->process_request($self, $calls);
+has _txn_level => (
+  is => 'rw',
+  isa => 'Int',
+  init_arg => undef,
+  default => 0,
+);
+
+has state => (
+  is => 'rw',
+  isa => 'Ix::AccountState',
+  lazy => 1,
+  builder => '_build_state',
+  predicate => '_has_state',
+  clearer   => '_clear_state',
+);
+
+sub _build_state ($self) {
+  Ix::AccountState->new({
+    context      => $self,
+    account_type => $self->account_type,
+    accountId    => $self->accountId,
+  });
 }
 
-sub state ($self) {
-  $self->root_context->state_for_account(
-    $self->account_type,
-    $self->accountId,
-  );
+# A wrapper around DBIC's txn_do. What this does is:
+#
+#  * increment our transaction depth
+#  * localize our copy of the pending states
+#  * Run the transaction
+#  * If it succeeds, we copy out the states that actually need changing
+#    and ship them up to our outer scope. This ensures that we don't
+#    record state changes in internal calls to ix_set if something fails
+#    somewhere along the way.
+#  * decrement our transaction depth. If depth reaches 0, that means we're
+#    at the start of the transaction and we need to commit state changes
+#
+# This should probably only be used by ix_set() and any calls it makes
+# internally that may try to bump state (generally, anything that may lead
+# to a nested ix_set).
+sub txn_do ($self, $code) {
+  if (
+       $self->_txn_level == 0
+    && $self->_has_state
+  ) {
+    # We should *NOT* have gotten any state information before starting
+    # a brand new transaction tree. If so, something is wrong.
+    require Carp;
+    Carp::confess("We already have state before starting a transaction?!");
+  }
+
+  # Start of a tree? Localize state so it goes away when we're done
+  if ($self->_txn_level == 0) {
+    local $self->{state} = $self->_build_state;
```

Comment:
This `local` seems like an obvious error.  Probably this should be causing a failing test.